### ISSUE TITLE
kernels before 3.17 don't export ktime_divns()

### ISSUE THREAD

**Issue by thomasjfox** [ISDN4Linux/mISDN#2]
Hi Karsten,

mISDN git HEAD (currently 2edcc9ca23e800fd1dd0b8b2d728705a22c69b52) compiles fine with kernel 3.14.x.

Loading the mISDN_core module fails with "**undefined symbol ktime_divns**" though.

I've tracked it down: The EXPORT_SYMBOL_GPL in kernel/hrtimer.c (or kernel/time/hrtimer.c) was introduced in kernel 3.17 only with upstream commit 166afb64511eef08e13331b970c44fe91cea45ef ("ktime: Sanitize ktime_to_us/ms conversion")

Therefore I think all kernels before 3.17 need the Revert_693a012bf96b18b00ee4da2e027b11da11b1f9f4.patch file applied?

Personally I'll just patch my 3.14 kernel to add EXPORT_SYMBOL_GPL(ktime_divns),
that's what some -stable updates were doing anyway:
http://www.mail-archive.com/linux-kernel@vger.kernel.org/msg1189362.html

Cheers,
Thomas

**Comment by thomasjfox** [comment]
Special case though: ktime_divns() is only defined as a function if BITS_PER_LONG is < 64.

Code from include/kernel/hrtimer.h:
```
#if BITS_PER_LONG < 64
extern u64 ktime_divns(const ktime_t kt, s64 div);
#else /* BITS_PER_LONG < 64 */
# define ktime_divns(kt, div)           (u64)((kt).tv64 / (div))
#endif
```

**Comment by kkeil** [comment]
Yes I only tested compiling for some old kernels < 4.0 , not tried to load the modules.

**Comment by thomasjfox** [comment]
On Friday, 28 July 2017 07:00:36 CEST Karsten Keil wrote:
> Yes I only tested compiling for some old kernels < 4.0 , not tried to load
> the modules.

these messages appear on kernel 3.14.x, too:

WARNING: "isdnhdlc_decode" [mISDN-2.0.36/standalone/drivers/isdn/hardware/mISDN/netjet.ko] undefined!
WARNING: "isdnhdlc_encode" [mISDN-2.0.36/standalone/drivers/isdn/hardware/mISDN/netjet.ko] undefined!
WARNING: "isdnhdlc_out_init" [mISDN-2.0.36/standalone/drivers/isdn/hardware/mISDN/netjet.ko] undefined!
WARNING: "isdnhdlc_rcv_init" [mISDN-2.0.36/standalone/drivers/isdn/hardware/mISDN/netjet.ko] undefined!


I've seen this message for at least a few years now and nobody complained.
The netjet cards are probably not in use anymore, I guess.
